util/compression_test.cc

```diff
@@ -0,0 +1,204 @@
+ public:
+  explicit AutoSkipTestFlushBlockPolicyFactory(
+      const int window, std::shared_ptr<Statistics> statistics)
+      : window_(window), statistics_(statistics) {}
+
+  virtual const char* Name() const override {
+    return "AutoSkipTestFlushBlockPolicyFactory";
+  }
+
+  virtual FlushBlockPolicy* NewFlushBlockPolicy(
+      const BlockBasedTableOptions& /*table_options*/,
+      const BlockBuilder& data_block_builder) const override {
+    (void)data_block_builder;
+    return new AutoSkipTestFlushBlockPolicy(window_, data_block_builder,
+                                            statistics_);
+  }
+
+ private:
+  int window_;
+  std::shared_ptr<Statistics> statistics_;
+};
+
+class DBAutoSkip : public DBTestBase {
+ public:
+  Options options;
+  Random rnd_;
+  int key_cursor_ = 0;
+  DBAutoSkip()
+      : DBTestBase("db_auto_skip", /*env_do_fsync=*/true),
+        options(CurrentOptions()),
+        rnd_(231) {
+    options.compression_manager =
+        CreateAutoSkipCompressionManager(GetDefaultBuiltinCompressionManager());
+    auto statistics = ROCKSDB_NAMESPACE::CreateDBStatistics();
+    options.statistics = statistics;
+    options.statistics->set_stats_level(StatsLevel::kExceptTimeForMutex);
+    BlockBasedTableOptions bbto;
+    bbto.enable_index_compression = false;
+    bbto.flush_block_policy_factory.reset(
+        new AutoSkipTestFlushBlockPolicyFactory(10, statistics));
+    options.table_factory.reset(NewBlockBasedTableFactory(bbto));
+    DestroyAndReopen(options);
+  }
+
+  bool CompressionFriendlyPut(const int no_of_kvs, const int size_of_value) {
+    auto value = std::string(size_of_value, 'A');
+    for (int i = 0; i < no_of_kvs; ++i) {
+      auto status = Put(Key(key_cursor_), value);
+      EXPECT_EQ(status.ok(), true);
+      key_cursor_++;
+    }
+    return true;
+  }
+  bool CompressionUnfriendlyPut(const int no_of_kvs, const int size_of_value) {
+    auto value = rnd_.RandomBinaryString(size_of_value);
+    for (int i = 0; i < no_of_kvs; ++i) {
+      auto status = Put(Key(key_cursor_), value);
+      EXPECT_EQ(status.ok(), true);
+      key_cursor_++;
+    }
+    return true;
+  }
+};
+
+// test case just to make sure auto compression manager is working
+TEST_F(DBAutoSkip, AutoSkipCompressionManager) {
+  if (GetSupportedCompressions().size() > 1) {
+    // AutoSkipCompressionManager starts with rejection ratio 0 i.e.
+    // compression enabled enough data to change the decision
+    const int kValueSize = 20000;
+    // Set the rejection ratio to 60%
+    CompressionUnfriendlyPut(6, kValueSize);
+    CompressionFriendlyPut(4, kValueSize);
+    // All the keys should be bypassed
+    CompressionUnfriendlyPut(6, kValueSize);
+    CompressionFriendlyPut(4, kValueSize);
+    // Set rejection ratio to 40%
+    CompressionUnfriendlyPut(4, kValueSize);
+    CompressionFriendlyPut(6, kValueSize);
+    // Compression attempted 6 rejected 4 accepted
```

Comment:
// This will verify all the data block compression are attempted based on previous prediction